clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java

```diff
@@ -73,79 +71,127 @@ public class TopicMetadataRequestManager implements RequestManager {
+                retryBackoffMaxMs);
+        inflightRequests.add(newRequest);
+        return newRequest.future;
+    }
 
+    /**
+     * Return the future of the metadata request.
+     *
+     * @param topic to be requested.
+     * @return the future of the metadata request.
+     */
+    public CompletableFuture<Map<String, List<PartitionInfo>>> requestTopicMetadata(final String topic, final long expirationTimeMs) {
         TopicMetadataRequestState newRequest = new TopicMetadataRequestState(
                 logContext,
                 topic,
+                expirationTimeMs,
                 retryBackoffMs,
                 retryBackoffMaxMs);
-        inflightRequests.put(topic, newRequest);
+        inflightRequests.add(newRequest);
         return newRequest.future;
     }
 
     // Visible for testing
     List<TopicMetadataRequestState> inflightRequests() {
-        return new ArrayList<>(inflightRequests.values());
+        return inflightRequests;
     }
 
     class TopicMetadataRequestState extends RequestState {
-        private final Optional<String> topic;
+        private final String topic;
+        private final boolean allTopics;
+        private final long expirationTimeMs;
         CompletableFuture<Map<String, List<PartitionInfo>>> future;
 
         public TopicMetadataRequestState(final LogContext logContext,
-                                         final Optional<String> topic,
+                                         final long expirationTimeMs,
+                                         final long retryBackoffMs,
+                                         final long retryBackoffMaxMs) {
+            super(logContext, TopicMetadataRequestState.class.getSimpleName(), retryBackoffMs,
+                    retryBackoffMaxMs);
+            future = new CompletableFuture<>();
+            this.topic = null;
+            this.allTopics = true;
+            this.expirationTimeMs = expirationTimeMs;
+        }
+
+        public TopicMetadataRequestState(final LogContext logContext,
+                                         final String topic,
+                                         final long expirationTimeMs,
                                          final long retryBackoffMs,
                                          final long retryBackoffMaxMs) {
             super(logContext, TopicMetadataRequestState.class.getSimpleName(), retryBackoffMs,
                 retryBackoffMaxMs);
             future = new CompletableFuture<>();
             this.topic = topic;
+            this.allTopics = false;
+            this.expirationTimeMs = expirationTimeMs;
         }
 
         /**
          * prepare the metadata request and return an
          * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.UnsentRequest} if needed.
          */
         private Optional<NetworkClientDelegate.UnsentRequest> send(final long currentTimeMs) {
+            if (currentTimeMs >= expirationTimeMs) {
```

Comment:
Do we need to do this? This is that thing again, the original code always sends the request, even if the timeout is expired, because the timeout only refers to the amount of time we are blocking. I.e. we should execute all operations that are non-blocking.

Reply:
If the request waiting for the metadata will have given up waiting, there didn't seem to me to be any point in sending the request any longer.

Reply:
Ah, yes, the story is different here because the call doesn't really have an effect